File: packages/protocol/src/topics/colors.ts
/**
 * Topic `colors` — the colour table, the special indices, and ramps.
 * OWNER: WP-22.
 *
 * Every `color` field on every other topic (`ObjectRow.color`,
 * `IndexedMeshHeader.oneColor`, CGO colour arrays) is a PyMOL colour INDEX.
 * Without this table the client cannot render any of them, so this topic is a
 * hard dependency of the object panel and of Mode G.
 *
 * WHAT IS AND IS NOT A LOOKUP
 * ---------------------------
 * A PyMOL colour index is not simply an offset into a table. `ColorGetIndex`
 * (`packages/engine/layer1/Color.cpp:661-750`) hands back four *different kinds* of number and
 * the client has to be able to tell them apart before it can produce an RGB:
 *
 *   `0 .. 5387`     a real slot in `CColor::Color` (`packages/engine/layer1/Color.cpp:825-1322`)
 *   `-1 .. -7`      the seven keywords (`packages/engine/layer1/Color.h:36-44`)
 *   `<= -10`        a **ramp object**; `ext = -10 - index` (`Color.h:46`)
 *   `& 0xC0000000 == 0x40000000`  an INLINE RGB, 8 bits per channel, packed by
 *                   `Color3fToInt` (`Color.cpp:1883`) and by `spectrumany`
 *                   (`packages/engine/modules/pymol/viewing.py:2053`)
 *
 * The decoders below are the client half of that, and they are pure: the wire
 * carries indices, this module turns the two *self-describing* kinds (inline
 * RGB, ramp slot) into something renderable with no round trip, and names the
 * two that genuinely need the backend (table slots, and `front`/`back`, which
 * depend on `bg_rgb` — `ColorUpdateFront`, `Color.cpp:1754`).
 */

/* ------------------------------------------------------------------ *
 * The table
 * ------------------------------------------------------------------ */

export interface ColorEntry {
  index: number;
  name: string;
  /** 0..1 floats, as `cmd.get_color_tuple()` returns them. */
  rgb: readonly [number, number, number];
}

/** A named colour ramp and the object it lives on (`object:ramp`). */
export interface ColorRamp {
  name: string;
  /** Object name of the ramp (`object:ramp`). */
  object: string;
}

/** The colour table plus ramps, as a full snapshot or an incremental patch. */
export interface ColorsPayload {
  /** `cmd.get_color_indices()` enriched with `cmd.get_color_tuple()`. */
  colors: ColorEntry[];
  ramps: ColorRamp[];
  /** True when this replaces the whole table rather than patching it. */
  full: boolean;
}

/**
 * `ColorReset` registers 188 explicit + 5200 generated slots
 * (`packages/engine/layer1/Color.cpp:825-1322`). Measured against this tree:
 * `len(cmd.get_color_indices(all=1)) == 5388`.
 */
export const COLOR_TABLE_SIZE = 5388;

/**
 * `ColorGetStatus` returns 1 only for names with no digits
 * (`packages/engine/layer1/Color.cpp:784-807`), and `cmd.get_color_indices()` (mode 1) returns
 * exactly those. Measured: 178.
 */
export const NAMED_COLOR_COUNT = 178;

/**
 * Index landmarks the Qt menus hardcode. Kept here because they are the cheap
 * runtime proof that the table we fetched is the table PyMOL built — a shifted
 * table would silently mis-colour every object row.
 * (`_gui.py:456-461`, `:632`; `packages/engine/layer1/Color.cpp:35-75`.)
 */
export const COLOR_LANDMARKS: Readonly<Record<string, number>> = {
  white: 0,
  black: 1,
  grey50: 104,
  grey80: 134,
  lightmagenta: 154,
  gray80: 4236,
  deepteal: 5262,
  darksalmon: 5280,
};

/* ------------------------------------------------------------------ *
 * Special indices — packages/engine/layer1/Color.h:36-47
 * ------------------------------------------------------------------ */

/** Inherit the colour from the enclosing scope (unset). */
export const C_COLOR_DEFAULT = -1;
/** The next auto colour a new object would receive. */
export const C_COLOR_NEW_AUTO = -2;
/** The current auto colour. */
export const C_COLOR_CUR_AUTO = -3;
/** Per-atom colour. */
export const C_COLOR_ATOMIC = -4;
/** The object's own colour. */
export const C_COLOR_OBJECT = -5;
/** Contrast-with-background colour. */
export const C_COLOR_FRONT = -6;
/** The background colour. */
export const C_COLOR_BACK = -7;
/** Indices <= this are ramp objects. `packages/engine/layer1/Color.h:46`. */
export const C_COLOR_EXT_CUTOFF = -10;

/** High bit marking an index as an inline `0x40RRGGBB` colour. */
export const C_COLOR_TRGB_BITS = 0x40000000;
/** Mask isolating the inline-colour marker bits. */
export const C_COLOR_TRGB_MASK = 0xc0000000;

/** The seven reserved colour keywords `ColorGetIndex` matches exactly. */
export type SpecialColorKeyword =
  'default' | 'auto' | 'current' | 'atomic' | 'object' | 'front' | 'back';

/**
 * The seven keywords `ColorGetIndex` matches EXACTLY (prefix matching was
 * removed in 2.5 — `packages/engine/layer1/Color.cpp:715-729`).
 *
 * `constant` is false for `auto` and `current`: those two are not constants at
 * all. `ColorGetIndex` runs `ColorGetNext()` / `ColorGetCurrent()`
 * (`Color.cpp:140,156`) and returns a REAL table index that changes as objects
 * are created — measured on a fresh session, `cmd.get_color_index('auto')`
 * answered 26 and `'current'` answered 5, not -2 and -3. So the client must
 * never cache those two and must never assume their sign.
 */
export const SPECIAL_COLORS: readonly {
  keyword: SpecialColorKeyword;
  index: number;
  constant: boolean;
  help: string;
}[] = [
  { keyword: 'default', index: C_COLOR_DEFAULT, constant: true, help: 'inherit (unset)' },
  { keyword: 'auto', index: C_COLOR_NEW_AUTO, constant: false, help: 'next auto colour' },
  { keyword: 'current', index: C_COLOR_CUR_AUTO, constant: false, help: 'current auto colour' },
  { keyword: 'atomic', index: C_COLOR_ATOMIC, constant: true, help: 'per-atom colour' },
  { keyword: 'object', index: C_COLOR_OBJECT, constant: true, help: 'object colour' },
  { keyword: 'front', index: C_COLOR_FRONT, constant: true, help: 'contrast with background' },
  { keyword: 'back', index: C_COLOR_BACK, constant: true, help: 'background colour' },
];

/* ------------------------------------------------------------------ *
 * Decoders
 * ------------------------------------------------------------------ */

/** Is this an inline `0x40RRGGBB` colour rather than a table slot? */
export function isInlineColor(index: number): boolean {
  if (!Number.isFinite(index) || index < 0) return false;
  // Unsigned throughout: JS bitwise ops are signed 32-bit and the 0xC0000000
  // mask would otherwise compare as a negative number.
  return ((index >>> 0) & C_COLOR_TRGB_MASK) >>> 0 === C_COLOR_TRGB_BITS;
}

/** `0x40RRGGBB` -> 0..1 floats. Inverse of `Color3fToInt` (`Color.cpp:1883`). */
export function decodeInlineColor(index: number): [number, number, number] {
  const packed = index >>> 0;
  return [((packed >>> 16) & 0xff) / 255, ((packed >>> 8) & 0xff) / 255, (packed & 0xff) / 255];
}

/**
 * 0..1 floats -> `0x40RRGGBB`, exactly as `spectrumany` does it
 * (`packages/engine/modules/pymol/viewing.py:2053`: `0x40000000 + r*0x10000 + g*0x100 + b`).
 */
export function encodeInlineColor(rgb: readonly [number, number, number]): number {
  return (
    (C_COLOR_TRGB_BITS + byte255(rgb[0]) * 0x10000 + byte255(rgb[1]) * 0x100 + byte255(rgb[2])) >>>
    0
  );
}

/** Is this a ramp-object colour (`index <= -10`)? */
export function isRampColor(index: number): boolean {
  return index <= C_COLOR_EXT_CUTOFF;
}

/** The colour-extension slot behind a ramp index. `ext = -10 - index`. */
export function rampSlot(index: number): number {
  return C_COLOR_EXT_CUTOFF - index;
}

/** The distinct kinds of value a colour index can encode. */
export type ColorKind = 'slot' | 'inline' | 'ramp' | 'special' | 'invalid';

/** Which of the four kinds of number is this? */
export function colorKind(index: number): ColorKind {
  if (!Number.isFinite(index) || !Number.isInteger(index)) return 'invalid';
  if (isInlineColor(index)) return 'inline';
  if (index >= 0) return index < COLOR_TABLE_SIZE ? 'slot' : 'invalid';
  if (index <= C_COLOR_EXT_CUTOFF) return 'ramp';
  if (index >= C_COLOR_BACK) return 'special';
  return 'invalid';
}

/**
 * `ColorGetName` (`packages/engine/layer1/Color.cpp:759-782`) inverted for the one kind the
 * client can name with no backend at all. Returns null otherwise.
 */
export function inlineColorName(index: number): string | null {
  if (!isInlineColor(index)) return null;
  return '0x' + (((index >>> 0) & 0xffffff) >>> 0).toString(16).padStart(6, '0');
}

/* ------------------------------------------------------------------ *
 * RGB formatting — shared by the picker, the swatch grid and the ramps
 * ------------------------------------------------------------------ */

function byte255(v: number): number {
  return Math.max(0, Math.min(255, Math.round(v * 255)));
}

/** 0..1 floats -> `#rrggbb`, for CSS. */
export function rgbToCss(rgb: readonly [number, number, number]): string {
  return (
    '#' +
    byte255(rgb[0]).toString(16).padStart(2, '0') +
    byte255(rgb[1]).toString(16).padStart(2, '0') +
    byte255(rgb[2]).toString(16).padStart(2, '0')
  );
}

/** `#rrggbb` / `0xrrggbb` / `rrggbb` -> 0..1 floats, or null. */
export function cssToRgb(text: string): [number, number, number] | null {
  const hex = text.trim().replace(/^#/, '').replace(/^0x/i, '');
  if (!/^[0-9a-f]{6}$/i.test(hex)) return null;
  return [
    parseInt(hex.slice(0, 2), 16) / 255,
    parseInt(hex.slice(2, 4), 16) / 255,
    parseInt(hex.slice(4, 6), 16) / 255,
  ];
}

/**
 * PyMOL menu labels embed one-digit-per-channel colour escapes — `\900` is
 * red, `\555` mid grey (`packages/engine/modules/pymol/menu.py:519-618`, drawn by
 * `packages/engine/layer1/Text.cpp`). The swatch grid's tile colours are those triples.
 */
export function menuSwatchToRgb(three: string): [number, number, number] {
  const digit = (i: number) => {
    const c = three.charCodeAt(i) - 48;
    return c >= 0 && c <= 9 ? c / 9 : 0;
  };
  return [digit(0), digit(1), digit(2)];
}

File: packages/bridge/tenmol_bridge/panels/colors.py
"""The colour service — the whole 5388-slot table in one engine-thread pass.

OWNER: WP-22.  Consumed by ``apps/web/src/features/colors``.

WHY THIS MODULE EXISTS
----------------------
There is no bulk colour accessor anywhere in PyMOL.  ``CmdGetColor``
(``packages/engine/layer4/Cmd.cpp:1321-1391``) has four modes and the only one that returns RGB
returns it for **one** colour::

    mode 0  name/index -> (r,g,b)                 one colour
    mode 1  -> [(name,index), ...]                the 178 digit-free names
    mode 2  -> [(name,index), ...]                all 5388 slots, NO rgb
    mode 3  name -> index
    mode 4  name/index -> (r,g,b) with a NEGATIVE r for the specials

So "the palette" is 1 call for the names plus 5388 calls for the values.  On
the engine thread that whole loop is **3.1 ms** (measured on this tree,
``SingletonPyMOL``, no GL work); the point of doing it here rather than in the
browser is not that it is impossible over the wire — 5388 pipelined WebSocket
calls measured 184 ms against the real bridge — but that it is one round trip
instead of 5388, and that ``cmd.space()`` invalidates the entire cache, so the
refetch happens more than once per session.

REACHABILITY (read this before assuming a client can call it)
-------------------------------------------------------------
``Dispatcher.resolve`` maps a wire ``fn`` root to ``pymol.<root>``
(``packages/bridge/tenmol_bridge/dispatch.py:_ROOT_MODULES``) and ``policy/base.py``
gates the root against ``DEFAULT_ROOTS``.  Neither knows about
``tenmol_bridge.panels``, and ``panels/__init__.py`` is a **frozen barrel**
whose ``PANELS`` table lists only ``objects/movie/seqview/menus``.  Wiring this
module up therefore needs two edits in files WP-22 does not own — a
``_bridge.colors.*`` branch in ``BridgeServer.bridge_route``
(``server.py:169-190``) or a ``PANELS`` entry plus a grant.  Until then the web
client fetches the table with the plain ``cmd.*`` calls this module wraps, and
every function here is exercised against a live PyMOL by
``packages/bridge/tests/test_colors.py``.

Everything in this module must run ON THE ENGINE THREAD (it calls ``cmd``), and
nothing in it draws.
"""

from __future__ import annotations

from typing import Any, Dict, Iterable, List, Optional, Sequence, Tuple

__all__ = [
    "COLOR_TABLE_SIZE",
    "NAMED_COLOR_COUNT",
    "SPECIAL_KEYWORDS",
    "COLOR_LANDMARKS",
    "palette",
    "named_colors",
    "tuples",
    "specials",
    "ramps",
    "volume_ramp_names",
    "define",
    "spectrum",
    "spectrum_any",
    "space",
    "menu_tree",
    "snapshot",
]

#: ``ColorReset`` registers 188 explicit + 5200 generated slots
#: (``packages/engine/layer1/Color.cpp:825-1322``).  Asserted, not trusted.
COLOR_TABLE_SIZE = 5388

#: ``ColorGetStatus`` == 1 only for names with no digits
#: (``packages/engine/layer1/Color.cpp:784-807``); that is what mode 1 returns.
NAMED_COLOR_COUNT = 178

#: The seven words ``ColorGetIndex`` matches exactly (``Color.cpp:715-729``).
SPECIAL_KEYWORDS: Tuple[str, ...] = (
    "default",
    "auto",
    "current",
    "atomic",
    "object",
    "front",
    "back",
)

#: Indices the Qt menus hardcode (``_gui.py:456-461,632``; ``Color.cpp:35-75``).
#: A shifted table would silently mis-colour everything, so we check.
COLOR_LANDMARKS: Dict[str, int] = {
    "white": 0,
    "black": 1,
    "grey50": 104,
    "grey80": 134,
    "lightmagenta": 154,
    "gray80": 4236,
    "deepteal": 5262,
    "darksalmon": 5280,
}


# --------------------------------------------------------------------------- #
# the table
# --------------------------------------------------------------------------- #


def named_colors(cmd: Any) -> List[Tuple[str, int]]:
    """``cmd.get_color_indices()`` — the 178 digit-free names."""
    return [(str(n), int(i)) for (n, i) in cmd.get_color_indices()]


def palette(cmd: Any, all: bool = True) -> List[Dict[str, Any]]:
    """``[{index, name, rgb}]`` for every registered slot.

    ``all=False`` restricts to the 178 digit-free names, which is what the Qt
    colour editor's ``list_colors`` shows (``pymol_qt_gui.py:547-611``).

    The RGB comes from ``get_color_tuple`` and NOT from any client-side
    reconstruction of the generated bands, because ``cmd.space()`` rewrites
    ``ColorRec::LutColor`` (``packages/engine/layer1/Color.cpp:1680``) and ``get_color_tuple``
    returns the LUT-mapped value.  A formula would be right until the first
    ``space cmyk`` and then silently wrong.
    """
    pairs = cmd.get_color_indices(all=1) if all else cmd.get_color_indices()
    out: List[Dict[str, Any]] = []
    for name, index in pairs:
        rgb = cmd.get_color_tuple(int(index))
        out.append(
            {
                "index": int(index),
                "name": str(name),
                "rgb": [float(rgb[0]), float(rgb[1]), float(rgb[2])] if rgb else None,
            }
        )
    return out


def tuples(cmd: Any, indices: Iterable[Any], mode: int = 0) -> List[Optional[List[float]]]:
    """``get_color_tuple`` for many names/indices in one round trip.

    ``mode=4`` is ``ColorGetSpecial`` (``Color.cpp:1789``), which flags a
    special colour with a NEGATIVE red component instead of failing.
    """
    out: List[Optional[List[float]]] = []
    for item in indices:
        rgb = cmd.get_color_tuple(item, mode)
        out.append([float(rgb[0]), float(rgb[1]), float(rgb[2])] if rgb else None)
    return out


def specials(cmd: Any) -> List[Dict[str, Any]]:
    """The seven keywords, resolved live.

    ``auto`` and ``current`` are NOT constants: ``ColorGetIndex`` runs
    ``ColorGetNext``/``ColorGetCurrent`` (``Color.cpp:140,156``) and hands back
    a real table index that moves as objects are created.  Callers that assume
    -2/-3 are wrong; this is why the resolution happens per request.
    """
    out: List[Dict[str, Any]] = []
    for word in SPECIAL_KEYWORDS:
        index = int(cmd.get_color_index(word))
        rgb = cmd.get_color_tuple(index, 4)
        out.append(
            {
                "keyword": word,
                "index": index,
                "rgb": [float(rgb[0]), float(rgb[1]), float(rgb[2])] if rgb else None,
            }
        )
    return out


# --------------------------------------------------------------------------- #
# ramps
# --------------------------------------------------------------------------- #


def ramps(cmd: Any) -> List[Dict[str, Any]]:
    """Live ``object:ramp`` objects and their negative colour index.

    Same query ``pymol.menu``'s ``menucontext`` uses (``menu.py:33-36``): there
    is no ``cmd.get_ramps()``.  ``index`` is the ``-10 - ext_slot`` colour a
    ramp can be used AS (``packages/engine/layer1/Color.h:46``).
    """
    out: List[Dict[str, Any]] = []
    for name in cmd.get_names("objects"):
        if cmd.get_type(name) != "object:ramp":
            continue
        try:
            index = int(cmd.get_color_index(name))
        except Exception:  # noqa: BLE001 - a ramp with no ext slot is still a ramp
            index = 0
        out.append({"name": str(name), "object": str(name), "index": index})
    return out


def volume_ramp_names() -> List[str]:
    """``pymol.colorramping.namedramps`` — the volume presets (``colorramping.py:17-54``)."""
    from pymol.colorramping import namedramps

    return sorted(namedramps)


# --------------------------------------------------------------------------- #
# writes
# --------------------------------------------------------------------------- #


def define(
    cmd: Any,
    name: str,
    rgb: Sequence[float],
    recolor: bool = True,
) -> Dict[str, Any]:
    """``set_color`` + ``recolor``, always in 0..1 floats.

    ``cmd.set_color`` auto-detects the range with ``if any component > 1.0:
    divide by 255`` (``viewing.py:2205-2207``), which means ``[1, 1, 1]`` is
    white and ``[1, 1, 2]`` is nearly black.  Sending floats removes the trap.
    ``recolor`` is required or existing objects keep the old RGB
    (``viewing.py:1868``, and the Qt editor issues it too —
    ``pymol_qt_gui.py:604``).
    """
    values = [float(c) for c in rgb]
    if len(values) != 3:
        raise ValueError("rgb must be 3 floats in 0..1")
    values = [min(1.0, max(0.0, c)) for c in values]
    cmd.set_color(str(name), values)
    if recolor:
        cmd.recolor()
    index = int(cmd.get_color_index(str(name)))
    tup = cmd.get_color_tuple(index)
    return {
        "name": str(name),
        "index": index,
        "rgb": [float(tup[0]), float(tup[1]), float(tup[2])] if tup else values,
    }


def spectrum(
    cmd: Any,
    expression: str = "count",
    palette_name: str = "rainbow",
    selection: str = "(all)",
    minimum: Optional[float] = None,
    maximum: Optional[float] = None,
    byres: int = 0,
    quiet: int = 1,
    interpolation: str = "rgb",
) -> Dict[str, Any]:
    """``cmd.spectrum`` (``viewing.py:2065-2151``); returns ``(min, max)``.

    ``minimum``/``maximum`` are left as ``None`` for auto-ranging — the C call
    signals that with ``minimum=0, maximum=-1`` (``viewing.py:2143-2145``) and
    the Python wrapper does that translation itself.

    NOTE the fallback: a non-alphabetic expression or an unknown palette makes
    ``cmd.spectrum`` delegate to ``spectrumany`` (``viewing.py:2137-2140``), so
    the palette argument is not validated here — PyMOL's own resolution order
    is the contract.
    """
    result = cmd.spectrum(
        expression,
        palette_name,
        selection,
        minimum,
        maximum,
        byres=int(byres),
        quiet=int(quiet),
        interpolation=interpolation,
    )
    return _range_result(result)


def spectrum_any(
    cmd: Any,
    expression: str,
    colors: str,
    selection: str = "(all)",
    minimum: Optional[float] = None,
    maximum: Optional[float] = None,
    quiet: int = 1,
    interpolation: str = "rgb",
) -> Dict[str, Any]:
    """``spectrumany`` (``viewing.py:1978-2063``) — arbitrary colour lists.

    MEASURED CORRECTION to the inventory's "``cmd.spectrumany``": there is no
    such attribute.  ``spectrumany`` is a module-level function in
    ``viewing.py`` that is never bound onto ``cmd`` (grep: it appears at
    ``viewing.py:1978`` and at ``viewing.py:2134`` and nowhere else), so
    ``cmd.spectrumany(...)`` raises ``AttributeError`` — verified against this
    tree.  The ONLY way in is ``cmd.spectrum`` with a palette argument that
    ``palette_sc`` cannot resolve: ``viewing.py:2130-2135`` then delegates the
    whole call.  A space-separated colour list is exactly that.

    ``colors`` is a space-separated list of colour names.  Interpolation is
    ``rgb`` (default), ``hls`` or ``hsv``
    (``_spectrumany_interpolations``, ``viewing.py:1968-1973``).

    Note that ``byres`` is deliberately absent: the fallback at
    ``viewing.py:2134`` does not forward it, so offering it would be a lie.
    """
    result = cmd.spectrum(
        expression,
        colors,
        selection,
        minimum,
        maximum,
        quiet=int(quiet),
        interpolation=interpolation,
    )
    return _range_result(result)


def _range_result(result: Any) -> Dict[str, Any]:
    if isinstance(result, (list, tuple)) and len(result) == 2:
        return {"minimum": float(result[0]), "maximum": float(result[1])}
    return {"minimum": None, "maximum": None}


def space(cmd: Any, space_name: str = "rgb", gamma: float = 1.0) -> Dict[str, Any]:
    """``cmd.space`` (``importing.py:227-288``) + the palette it invalidates.

    ``ColorRec::LutColor`` (``packages/engine/layer1/Color.h:58-59``) means every colour's RGB
    can change, so the answer carries the fresh table rather than leaving the
    client to guess that it must refetch.
    """
    cmd.space(space_name, gamma)
    return {"space": space_name, "gamma": float(gamma), "colors": palette(cmd, all=True)}


# --------------------------------------------------------------------------- #
# the "C" menu
# --------------------------------------------------------------------------- #

#: ``pymol.menu`` providers that produce a colour tree, by the popup name the
#: internal GUI uses (``packages/engine/modules/pymol/menu.py:643-712``).
_MENU_PROVIDERS = (
    "mol_color",
    "measurement_color",
    "general_color",
    "slice_color",
    "vol_color",
)


def menu_tree(cmd: Any, sele: str = "(all)", kind: str = "mol_color") -> List[Any]:
    """PyMOL's OWN colour menu, normalised to JSON.

    ``pymol.menu.mol_color(self_cmd, sele)`` (``menu.py:672-686``) takes the
    cmd instance as its first positional argument, so a client cannot call it
    over the wire — this is the wrapper that supplies it.

    Items are ``[type, label, action]`` with type 1 = item/submenu, 2 = title,
    0 = separator (``packages/engine/layer4/PopUp.cpp``), and ``action`` is either a command
    STRING or a nested list.  Both shapes survive here unchanged; labels keep
    their ``\\RGB`` inline colour escapes.
    """
    if kind not in _MENU_PROVIDERS:
        raise ValueError(
            "unknown colour menu %r; known: %s" % (kind, ", ".join(_MENU_PROVIDERS))
        )
    import pymol.menu as pymol_menu

    provider = getattr(pymol_menu, kind)
    return _normalise(provider(cmd, sele))


def _normalise(items: Any) -> Any:
    """Menu lists are tuples/lists of mixed types; make them JSON-safe."""
    if isinstance(items, (list, tuple)):
        return [_normalise(item) for item in items]
    if isinstance(items, (str, int, float, bool)) or items is None:
        return items
    return str(items)


# --------------------------------------------------------------------------- #
# the topic payload
# --------------------------------------------------------------------------- #


def snapshot(cmd: Any, all: bool = True) -> Dict[str, Any]:
    """A ``ColorsPayload`` (``packages/protocol/src/topics/colors.ts``)."""
    return {
        "colors": palette(cmd, all=all),
        "ramps": [{"name": r["name"], "object": r["object"]} for r in ramps(cmd)],
        "full": bool(all),
    }
